### DISCUSSION TITLE
Apple Should Buy This

### DISCUSSION THREAD

**Discussion by DrDauger** [hishnash/KeyWindow#2] [Q&A]
I'm about to test it shortly, but this KeyWindow implementation looks really good and thorough.  It fills in a major gap in SwiftUI 2 on macOS (& iPadOS).  

Is there any indication Apple will put in a fix anytime soon?  If not, those of us writing for SwiftUI will have to rely on something like this.  

P.S. Makes me long for the Inside Mac Volume I days when you could just wait for activateEvt to know which window was key then call the Menu Manager to manipulate the menus....

**Comment by hishnash** [comment]
If the focus api is updated to properly support macOS and iOS, in the same way as NSResponder so that the active window gets `focus` then this package will no longer be needed. Currently the focus api is clearly only 10% finished with its only useable platform being tvOS. 

Also worth noting that while this package works on iPadOS if your hoping to use it to get global keyboard shortcuts even through this was demoed in WWDC.. it seems it never landed into a public (or beta) build of the OS. So the use case on iPadOS right now is limited to being able to have some form of connection between multiple scenes tracking the one that is `key`, eg in split view or slipover.

**Reply by DrDauger** [reply]
Well in any case we appreciate this solution.  Until WWDC21 or 22, I think that's a big "if". The idea of SwiftUI is great, but I thought Apple could do better with the follow through. I've found SwiftUI 2 to have a few other gaps and several glitchy behaviors. I can give SwiftUI 1 a pass, but Apple a year later still missing things the Mac had in the 1980s is slowing adoption of SwiftUI.  I'm spending a lot of development time writing other workarounds to cover up the problems.

**Reply by hishnash** [reply]
To be honist the complexity of what they are doing going from the UIKit/AppKit apis to a declarative apis is a massive task, (doing this is such a way as to avoid the envitable massive perfomance hits). 

One could have expected apple to keep these things in house for longer but on some platforms (the apple watch) swiftUI is so so so much better than the alternative that releasing it even the v1 version made lots of sense. 

On macOS I have my own list of wishes.. the issue solved in this package is not even close to the top of my wishist to be honest.
